src/happy.rs

```diff
@@ -26,73 +25,64 @@ where
         return TcpStream::connect_timeout(addr, timeout);
     }
 
-    let ipv4 = addrs.iter().filter(|a| a.is_ipv4()).copied();
-    let ipv6 = addrs.iter().filter(|a| a.is_ipv6()).copied();
+    let ipv4 = addrs.iter().filter(|a| a.is_ipv4());
+    let ipv6 = addrs.iter().filter(|a| a.is_ipv6());
     let sorted = intertwine(ipv6, ipv4);
 
     let (tx, rx) = channel();
     let mut first_err = None;
 
     let start = Instant::now();
 
+    let mut handle_res = |addr, res| match res {
+        Ok(sock) => {
+            debug!(
+                "successfully connected to {}, took {}ms",
+                addr,
+                start.elapsed().as_millis()
+            );
+
+            Some(sock)
+        }
+        Err(err) => {
+            debug!("failed to connect to {}: {}", addr, err);
+
+            if first_err.is_none() {
+                first_err = Some(err);
+            }
+
+            None
+        }
+    };
+
     // This loop will race each connection attempt against others, returning early if a
     // connection attempt is successful.
-    for addr in sorted {
+    for &addr in sorted {
         let tx = tx.clone();
 
         thread::spawn(move || {
             debug!("trying to connect to {}", addr);
 
-            let _ = tx.send(TcpStream::connect_timeout(&addr, timeout));
+            let _ = tx.send((addr, TcpStream::connect_timeout(&addr, timeout)));
         });
 
-        match rx.recv_timeout(RACE_DELAY) {
-            Ok(Ok(sock)) => {
-                debug!("success, took {}ms", start.elapsed().as_millis());
-
+        if let Ok((addr, res)) = rx.recv_timeout(RACE_DELAY) {
+            if let Some(sock) = handle_res(addr, res) {
                 return Ok(sock);
             }
-            Ok(Err(err)) => {
-                debug!("connection error: {} addr={}", err, addr);
-
-                if first_err.is_none() {
-                    first_err = Some(err);
-                }
-            }
-            Err(RecvTimeoutError::Timeout) => {
-                continue;
-            }
-            Err(RecvTimeoutError::Disconnected) => {
-                unreachable!();
-            }
         }
     }
 
     // We must drop this handle to the sender in order to properly disconnect the channel
     // when all the threads are finished.
     drop(tx);
 
-    // This loop waits for replies from the background threads. It will automatically timeout when the background
-    // thread's connection attempt timeouts and the senders are dropped.
+    // This loop waits for replies from the background threads. It will automatically timeout
+    // when the background thread's connection attempt timeouts and the senders are dropped.
```

Comment:
Can you fix the typo I made earlier? `when the background threads' connection attempts timeout`

Reply:
Done

Reply:
Thanks, merging.